# Stage 2: Vector Search + LLM (Retrieval-Augmented Generation)

In Stage 1, we saw the LLM confidently invent citations that don't exist.
To fix this, we will **give it real papers to reference.** This is precisely Retrieval-Augmented Generation (or RAG): retrieve first, then generate.

In this notebook we'll:
1. Convert PubMed abstracts into vector embeddings
2. Store them in Qdrant (a vector database)
3. Retrieve relevant papers for a question
4. Feed those papers to the LLM as context
5. Discover the limits of pure vector search

## 1. Setup + Load Data

In [ ]:
import os
import json

from dotenv import load_dotenv
from openai import OpenAI

from qdrant_client import QdrantClient
from qdrant_client.models import models

load_dotenv("../.env")

client = OpenAI()
qdrant = QdrantClient(url=os.getenv("QDRANT_URL", "http://localhost:6333"))

COLLECTION = "pubmed_dense"
EMBED_MODEL = "text-embedding-3-small"
EMBED_DIM = 1536

with open("../data/pubmed_dataset.json") as f:
    dataset = json.load(f)

# filter papers with abstracts, limit to 5,000 for demo
papers = [p for p in dataset["papers"] if p.get("abstract")][:5_000] 
print(f"Loaded {len(papers)} papers with abstracts")

Loaded 9762 papers with abstracts


## 2. What Are Vector Embeddings?

An embedding model converts text into a high-dimensional vector (a list of
numbers) where **similar meanings are close together** in the vector space.
OpenAI's `text-embedding-3-small` maps any text to a **1536-dimensional
vector**. Two abstracts about the same disease will end up near each other;
an abstract about oncology and one about baking will be far apart.

This is the core idea behind semantic search: instead of matching keywords,
we match *meaning*.

## 3. Create Qdrant Collection

In [11]:
# Delete if exists, create fresh
if qdrant.collection_exists(COLLECTION):
    qdrant.delete_collection(COLLECTION)

qdrant.create_collection(
    collection_name=COLLECTION,
    vectors_config=models.VectorParams(
        size=EMBED_DIM,
        distance=models.Distance.COSINE,
    ),
)
print(f"Collection '{COLLECTION}' created")

Collection 'pubmed_dense' created


## 4. Embed and Ingest Papers

We'll send abstracts to the OpenAI embedding API in batches of 100, then
upsert the resulting vectors (along with paper metadata) into Qdrant.

In [13]:
import time

from tqdm import tqdm

BATCH_SIZE = 100
MAX_TOKENS = 8191  # for text-embedding-3-small
start = time.time()

for i in tqdm(range(0, len(papers), BATCH_SIZE), desc="Ingesting papers"):
    batch = papers[i : i + BATCH_SIZE]
    texts = [p["abstract"][:MAX_TOKENS] for p in batch]

    # batch embed
    response = client.embeddings.create(model=EMBED_MODEL, input=texts)
    vectors = [e.embedding for e in response.data]

    # create points
    points = [
        models.PointStruct(
            id=int(paper["pmid"]),
            vector=vector,
            payload={
                "pmid": paper["pmid"],
                "title": paper["title"],
                "abstract": paper["abstract"][:MAX_TOKENS],
                "authors": [a["name"] for a in paper.get("authors", [])],
                "journal": paper.get("journal", ""),
                "mesh_terms": [m["term"] for m in paper.get("mesh_terms", [])],
            },
        )
        for paper, vector in zip(batch, vectors)
    ]

    qdrant.upsert(collection_name=COLLECTION, points=points)
    print(f"  Ingested {min(i + BATCH_SIZE, len(papers))}/{len(papers)} papers")

elapsed = time.time() - start
print(f"\nDone! {len(papers)} papers ingested in {elapsed:.1f}s")

Ingesting papers:   1%|          | 1/98 [00:01<01:58,  1.22s/it]

  Ingested 100/9762 papers


Ingesting papers:   2%|▏         | 2/98 [00:02<02:24,  1.51s/it]

  Ingested 200/9762 papers


Ingesting papers:   3%|▎         | 3/98 [00:03<02:02,  1.29s/it]

  Ingested 300/9762 papers


Ingesting papers:   4%|▍         | 4/98 [00:05<01:59,  1.27s/it]

  Ingested 400/9762 papers


Ingesting papers:   5%|▌         | 5/98 [00:06<01:57,  1.26s/it]

  Ingested 500/9762 papers


Ingesting papers:   6%|▌         | 6/98 [00:07<02:02,  1.33s/it]

  Ingested 600/9762 papers


Ingesting papers:   7%|▋         | 7/98 [00:09<02:08,  1.42s/it]

  Ingested 700/9762 papers


Ingesting papers:   8%|▊         | 8/98 [00:11<02:11,  1.46s/it]

  Ingested 800/9762 papers


Ingesting papers:   9%|▉         | 9/98 [00:11<01:54,  1.28s/it]

  Ingested 900/9762 papers


Ingesting papers:  10%|█         | 10/98 [00:14<02:26,  1.67s/it]

  Ingested 1000/9762 papers


Ingesting papers:  11%|█         | 11/98 [00:15<02:17,  1.59s/it]

  Ingested 1100/9762 papers


Ingesting papers:  12%|█▏        | 12/98 [00:16<02:01,  1.42s/it]

  Ingested 1200/9762 papers


Ingesting papers:  13%|█▎        | 13/98 [00:17<01:46,  1.25s/it]

  Ingested 1300/9762 papers


Ingesting papers:  14%|█▍        | 14/98 [00:19<01:52,  1.34s/it]

  Ingested 1400/9762 papers


Ingesting papers:  15%|█▌        | 15/98 [00:20<01:39,  1.20s/it]

  Ingested 1500/9762 papers


Ingesting papers:  16%|█▋        | 16/98 [00:21<01:46,  1.30s/it]

  Ingested 1600/9762 papers


Ingesting papers:  17%|█▋        | 17/98 [00:23<01:46,  1.32s/it]

  Ingested 1700/9762 papers


Ingesting papers:  18%|█▊        | 18/98 [00:23<01:33,  1.17s/it]

  Ingested 1800/9762 papers


Ingesting papers:  19%|█▉        | 19/98 [00:24<01:30,  1.15s/it]

  Ingested 1900/9762 papers


Ingesting papers:  20%|██        | 20/98 [00:26<01:32,  1.19s/it]

  Ingested 2000/9762 papers


Ingesting papers:  21%|██▏       | 21/98 [00:27<01:24,  1.10s/it]

  Ingested 2100/9762 papers


Ingesting papers:  22%|██▏       | 22/98 [00:28<01:20,  1.06s/it]

  Ingested 2200/9762 papers


Ingesting papers:  23%|██▎       | 23/98 [00:29<01:19,  1.06s/it]

  Ingested 2300/9762 papers


Ingesting papers:  24%|██▍       | 24/98 [00:29<01:08,  1.08it/s]

  Ingested 2400/9762 papers


Ingesting papers:  26%|██▌       | 25/98 [00:30<01:12,  1.00it/s]

  Ingested 2500/9762 papers


Ingesting papers:  27%|██▋       | 26/98 [00:31<01:09,  1.03it/s]

  Ingested 2600/9762 papers


Ingesting papers:  28%|██▊       | 27/98 [00:32<01:02,  1.14it/s]

  Ingested 2700/9762 papers


Ingesting papers:  29%|██▊       | 28/98 [00:33<01:00,  1.15it/s]

  Ingested 2800/9762 papers


Ingesting papers:  30%|██▉       | 29/98 [00:34<01:04,  1.06it/s]

  Ingested 2900/9762 papers


Ingesting papers:  31%|███       | 30/98 [00:35<01:11,  1.06s/it]

  Ingested 3000/9762 papers


Ingesting papers:  32%|███▏      | 31/98 [00:37<01:15,  1.12s/it]

  Ingested 3100/9762 papers


Ingesting papers:  33%|███▎      | 32/98 [00:37<01:07,  1.02s/it]

  Ingested 3200/9762 papers


Ingesting papers:  34%|███▎      | 33/98 [00:38<01:03,  1.02it/s]

  Ingested 3300/9762 papers


Ingesting papers:  35%|███▍      | 34/98 [00:39<01:04,  1.01s/it]

  Ingested 3400/9762 papers


Ingesting papers:  36%|███▌      | 35/98 [00:41<01:15,  1.19s/it]

  Ingested 3500/9762 papers


Ingesting papers:  37%|███▋      | 36/98 [00:42<01:19,  1.28s/it]

  Ingested 3600/9762 papers


Ingesting papers:  38%|███▊      | 37/98 [00:43<01:07,  1.10s/it]

  Ingested 3700/9762 papers


Ingesting papers:  39%|███▉      | 38/98 [00:44<01:02,  1.04s/it]

  Ingested 3800/9762 papers


Ingesting papers:  40%|███▉      | 39/98 [00:45<00:57,  1.03it/s]

  Ingested 3900/9762 papers


Ingesting papers:  41%|████      | 40/98 [00:46<00:53,  1.09it/s]

  Ingested 4000/9762 papers


Ingesting papers:  42%|████▏     | 41/98 [00:46<00:51,  1.11it/s]

  Ingested 4100/9762 papers


Ingesting papers:  43%|████▎     | 42/98 [00:47<00:48,  1.16it/s]

  Ingested 4200/9762 papers


Ingesting papers:  44%|████▍     | 43/98 [00:49<00:58,  1.07s/it]

  Ingested 4300/9762 papers


Ingesting papers:  45%|████▍     | 44/98 [00:50<00:54,  1.02s/it]

  Ingested 4400/9762 papers


Ingesting papers:  46%|████▌     | 45/98 [00:51<00:54,  1.02s/it]

  Ingested 4500/9762 papers


Ingesting papers:  47%|████▋     | 46/98 [00:52<00:54,  1.04s/it]

  Ingested 4600/9762 papers


Ingesting papers:  48%|████▊     | 47/98 [00:53<00:48,  1.05it/s]

  Ingested 4700/9762 papers


Ingesting papers:  49%|████▉     | 48/98 [00:54<00:59,  1.18s/it]

  Ingested 4800/9762 papers


Ingesting papers:  50%|█████     | 49/98 [00:55<00:53,  1.10s/it]

  Ingested 4900/9762 papers


Ingesting papers:  51%|█████     | 50/98 [00:56<00:51,  1.07s/it]

  Ingested 5000/9762 papers


Ingesting papers:  52%|█████▏    | 51/98 [00:57<00:49,  1.05s/it]

  Ingested 5100/9762 papers


Ingesting papers:  53%|█████▎    | 52/98 [00:58<00:44,  1.04it/s]

  Ingested 5200/9762 papers


Ingesting papers:  54%|█████▍    | 53/98 [00:59<00:43,  1.04it/s]

  Ingested 5300/9762 papers


Ingesting papers:  55%|█████▌    | 54/98 [01:00<00:42,  1.03it/s]

  Ingested 5400/9762 papers


Ingesting papers:  56%|█████▌    | 55/98 [01:01<00:41,  1.04it/s]

  Ingested 5500/9762 papers


Ingesting papers:  57%|█████▋    | 56/98 [01:02<00:44,  1.06s/it]

  Ingested 5600/9762 papers


Ingesting papers:  58%|█████▊    | 57/98 [01:05<01:00,  1.47s/it]

  Ingested 5700/9762 papers


Ingesting papers:  59%|█████▉    | 58/98 [01:07<01:07,  1.68s/it]

  Ingested 5800/9762 papers


Ingesting papers:  60%|██████    | 59/98 [01:08<01:04,  1.66s/it]

  Ingested 5900/9762 papers


Ingesting papers:  61%|██████    | 60/98 [01:11<01:10,  1.85s/it]

  Ingested 6000/9762 papers


Ingesting papers:  62%|██████▏   | 61/98 [01:11<00:55,  1.49s/it]

  Ingested 6100/9762 papers


Ingesting papers:  63%|██████▎   | 62/98 [01:13<00:53,  1.50s/it]

  Ingested 6200/9762 papers


Ingesting papers:  64%|██████▍   | 63/98 [01:14<00:45,  1.30s/it]

  Ingested 6300/9762 papers


Ingesting papers:  65%|██████▌   | 64/98 [01:14<00:38,  1.14s/it]

  Ingested 6400/9762 papers


Ingesting papers:  66%|██████▋   | 65/98 [01:15<00:35,  1.07s/it]

  Ingested 6500/9762 papers


Ingesting papers:  67%|██████▋   | 66/98 [01:18<00:51,  1.60s/it]

  Ingested 6600/9762 papers


Ingesting papers:  68%|██████▊   | 67/98 [01:20<00:52,  1.69s/it]

  Ingested 6700/9762 papers


Ingesting papers:  69%|██████▉   | 68/98 [01:23<00:57,  1.93s/it]

  Ingested 6800/9762 papers


Ingesting papers:  70%|███████   | 69/98 [01:24<00:53,  1.83s/it]

  Ingested 6900/9762 papers


Ingesting papers:  71%|███████▏  | 70/98 [01:27<01:02,  2.24s/it]

  Ingested 7000/9762 papers


Ingesting papers:  72%|███████▏  | 71/98 [01:30<01:01,  2.27s/it]

  Ingested 7100/9762 papers


Ingesting papers:  73%|███████▎  | 72/98 [01:31<00:52,  2.03s/it]

  Ingested 7200/9762 papers


Ingesting papers:  74%|███████▍  | 73/98 [01:34<00:53,  2.16s/it]

  Ingested 7300/9762 papers


Ingesting papers:  76%|███████▌  | 74/98 [01:36<00:56,  2.36s/it]

  Ingested 7400/9762 papers


Ingesting papers:  77%|███████▋  | 75/98 [01:38<00:49,  2.17s/it]

  Ingested 7500/9762 papers


Ingesting papers:  78%|███████▊  | 76/98 [01:41<00:49,  2.23s/it]

  Ingested 7600/9762 papers


Ingesting papers:  78%|███████▊  | 76/98 [01:42<00:29,  1.35s/it]


BadRequestError: Error code: 400 - {'error': {'message': "We could not parse the JSON body of your request. (HINT: This likely means you aren't using your HTTP library correctly. The OpenAI API expects a JSON payload, but what was sent was not valid JSON. If you have trouble figuring out how to fix this, please contact us through our help center at help.openai.com.)", 'type': 'invalid_request_error', 'param': None, 'code': None}}

## 5. Semantic Search

Now let's search! We embed our question into the same vector space and find
the nearest papers. Papers whose abstracts are semantically close to the
question will have the highest cosine similarity scores.

In [14]:
def search_papers(question: str, top_k: int = 5):
    """
    Embed the question and search Qdrant for similar papers.
    """
    q_embedding = client.embeddings.create(
        model=EMBED_MODEL, input=question
    ).data[0].embedding

    results = qdrant.query_points(
        collection_name=COLLECTION,
        query=q_embedding,
        limit=top_k,
        with_payload=True,
    )
    return results.points


# Search!
question = "What are the latest advances in using CRISPR-Cas9 for treating sickle cell disease?"
results = search_papers(question)

print(f"Query: {question}\n")
for i, point in enumerate(results, 1):
    print(f"{i}. [Score: {point.score:.4f}] {point.payload['title'][:90]}")
    print(f"   PMID: {point.payload['pmid']} | Journal: {point.payload['journal']}")
    print()

Query: What are the latest advances in using CRISPR-Cas9 for treating sickle cell disease?

1. [Score: 0.7530] Curing Sickle Cell Disease by Allogeneic Hematopoietic Stem Cell (HSC) Transplantation Tow
   PMID: 41300819 | Journal: Genes

2. [Score: 0.7441] Use of genome-editing tools to treat sickle cell disease.
   PMID: 27250347 | Journal: Human genetics

3. [Score: 0.7440] CRISPR/Cas9 System as a Promising Therapy in Thalassemia and Sickle Cell Disease: A System
   PMID: 39794549 | Journal: Molecular biotechnology

4. [Score: 0.7362] CRISPR/Cas9 in the treatment of sickle cell disease (SCD) and its comparison with traditio
   PMID: 39359808 | Journal: Annals of medicine and surgery (2012)

5. [Score: 0.7305] A justice-based argument for including sickle cell disease in CRISPR/Cas9 clinical researc
   PMID: 31107563 | Journal: Bioethics



## 6. RAG: Feed Context to the LLM

Here's the magic. Instead of asking the LLM to answer from its training
data (where it hallucinates), we **stuff the retrieved papers into the
prompt** and instruct it to answer using *only* that context.

In [15]:
def rag_answer(question: str, results: list[models.ScoredPoint]):
    """
    Generate an answer using retrieved papers as context.
    """
    context = "\n\n".join(
        [
            f"Paper {i+1} (PMID: {r.payload['pmid']}):\n"
            f"Title: {r.payload['title']}\n"
            f"Abstract: {r.payload['abstract'][:500]}"
            for i, r in enumerate(results)
        ]
    )

    response = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[
            {
                "role": "system",
                "content": (
                    "You are a biomedical research assistant. Answer questions "
                    "using ONLY the provided paper context. Cite papers by PMID."
                ),
            },
            {
                "role": "user",
                "content": f"Context:\n{context}\n\nQuestion: {question}",
            },
        ],
        temperature=0.0,
    )
    return response.choices[0].message.content


answer = rag_answer(question, results)
print(answer)

Recent advances in using CRISPR-Cas9 for treating sickle cell disease (SCD) include its application as a promising gene editing tool that can modify the genetic mutations responsible for the disease. Clinical trials are underway to evaluate the efficacy and safety of the CRISPR/Cas9 system in treating SCD, demonstrating its potential as a revolutionary approach compared to traditional treatment methods (PMID: 39794549, PMID: 39359808, PMID: 31107563). The technology allows for precise modifications of the β-globin gene, which is implicated in SCD, and has shown substantial success in preclinical and early clinical settings (PMID: 27250347).


## 7. RAG Works!

Now every citation is **real** -- the LLM is grounded in actual papers from
our database. No more hallucinated PMIDs. This is a massive improvement
over vanilla prompting.

But let's try a more specific, multi-faceted query and see what happens...

## 8. The Semantic Drift Problem

In [17]:
specific_question = "What papers discuss p53 tumor suppressor gene mutations specifically in glioblastoma multiforme?"
results_specific = search_papers(specific_question)

print(f"Query: {specific_question}\n")
for i, point in enumerate(results_specific, 1):
    print(f"{i}. [Score: {point.score:.4f}] {point.payload['title'][:90]}")
    # abstract sample
    print(f"   Abstract: {point.payload['abstract'][:200]}...")
    mesh = point.payload.get("mesh_terms", [])[:5]
    print(f"   MeSH: {', '.join(mesh)}")
    print()

Query: What papers discuss p53 tumor suppressor gene mutations specifically in glioblastoma multiforme?

1. [Score: 0.5536] Genome-wide CRISPR screens identify novel regulators of wild-type and mutant p53 stability
   Abstract: Tumor suppressor p53 (TP53) is frequently mutated in cancer, often resulting not only in loss of its tumor-suppressive function but also acquisition of dominant-negative and even oncogenic gain-of-fun...
   MeSH: Tumor Suppressor Protein p53, Animals, Humans, Mice, Protein Stability

2. [Score: 0.5518] Unlocking glioblastoma: breakthroughs in molecular mechanisms and next-generation therapie
   Abstract: Glioblastoma (GB) remains the most aggressive primary brain tumor in adults, characterized by rapid progression, recurrence, and resistance to conventional therapies. Despite advancements in surgical ...
   MeSH: Humans, Glioblastoma, Brain Neoplasms, Immunotherapy, Molecular Targeted Therapy

3. [Score: 0.5505] In vivo Engineering of Chromosome 19 q-arm by Empl

## 9. The Gap
Even with a partial index, dense retrieval worked well for broad queries (e.g., CRISPR + sickle cell disease), returning clearly relevant papers with strong scores (~0.73–0.75).

But the failure mode shows up on specific queries:

> *“What papers discuss p53 tumor suppressor gene mutations specifically in glioblastoma multiforme?”*

Top results had lower scores (~0.53–0.55) and mixed relevance:
- some papers were about **p53** but not specifically GBM,
- others were about **glioblastoma** but not focused on **p53 mutation evidence**.

**The problem:** dense vector search optimizes semantic similarity, not exact term matching.  
**What we need next:** combine semantic retrieval with lexical precision (keyword/BM25) — i.e., **hybrid search** in Stage 3.

## Architecture

```
Stage 2 architecture:
    Question --embed--> [Qdrant: Dense Vectors] --top-k--> Papers --> [LLM] --> Answer

Problem: Semantic drift -- related but wrong papers

Next: Add keyword search (BM25) for precision
```